### PR TITLE
Add `info` command

### PR DESCRIPTION
**Pull request by lean-apple**
Closes https://github.com/paritytech/cargo-contract/issues/783

###  Done : 

- [x] Decode Result from API Storage Contract Info of 
- [x] Handing if no contract/AccountId is provided - Add a check before making the request ? 
- [x] Format the result 
- [x] Add Integration Test 
- [x] Add documentation for the new command

### CODE REVIEW

**Comment by ascjones** [comment]
> Current display of the infos, maybe it can be still improved Screenshot 2023-03-06 at 11 06 34

The trie id should also be displayed as a hex encoded string, as byte strings usually are for display purposes

File: README.md [hunk 1]
```diff
@@ -121,6 +121,10 @@ The argument has to be given as hex-encoding, starting with `0x`.
 
 Remove a contract from a `pallet-contracts` enabled chain. See [extrinsics](docs/extrinsics.md).
 
+##### `cargo contract info`
+
+Fetch and display contract information of a contract on chain. See [extrinsics](docs/extrinsics.md).
```
Comment (ascjones) [inline README.md:126]: `info` is not an extrinsic.

File: crates/cargo-contract/src/cmd/extrinsics/info.rs [hunk 2]
```diff
@@ -0,0 +1,125 @@
+// Copyright 2018-2023 Parity Technologies (UK) Ltd.
```
Comment (ascjones) [inline crates/cargo-contract/src/cmd/extrinsics/info.rs:1]: This command should be moved out of the `extrinsics` folder

File: crates/cargo-contract/src/cmd/extrinsics/info.rs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                }
+                None => {
+                    return Err(anyhow!(
+                        "No contract information was found for the ContractId {}",
+                        self.contract
+                    )
+                    .into())
+                }
+            }
+            Result::<(), ErrorVariant>::Ok(())
```
Comment (ascjones) [inline crates/cargo-contract/src/cmd/extrinsics/info.rs:85]: Is this type hint necessary?

File: crates/cargo-contract/src/cmd/extrinsics/info.rs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        })
+    }
+
+    async fn info_dry_run(&self, client: &Client) -> Result<Option<ContractInfo>> {
```
Comment (ascjones) [inline crates/cargo-contract/src/cmd/extrinsics/info.rs:89]: `dry_run` applies to calling extrinsics over RPC. So this can be renamed to `fetch_contract_info` or similar.

File: docs/extrinsics.md [hunk 5]
```diff
@@ -1,5 +1,5 @@
 # Extrinsics
-`cargo-contract` provides CLI support for uploading, instantiating and calling your contracts directly from the command
+`cargo-contract` provides CLI support for uploading, instantiating, calling, removing your contracts and displaying info directly from the command
```
Comment (ascjones) [inline docs/extrinsics.md:2]: Remove the part about displaying info, since not an extrinsic.

File: docs/extrinsics.md [hunk 6]
```diff
@@ -116,6 +116,27 @@ cargo contract remove \
 - `--code-hash` the hash of the uploaded code, returned from a call to `contract upload`. 
 If not specified the code hash will be taken from the contract artifacts. 
 
+### `info`
```
Comment (ascjones) [inline docs/extrinsics.md:119]: Move this docs up out of `extrinsics.md`

File: crates/cargo-contract/src/cmd/extrinsics/info.rs (hunk 2 above)
Comment (lean-apple) [reply crates/cargo-contract/src/cmd/extrinsics/info.rs:1]: It is already in progress

File: crates/cargo-contract/src/cmd/extrinsics/info.rs [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+            .await?
+            .fetch(&info_contract_call)
+            .await?;
+
+        Ok(contract_info_of)
+    }
+
+    fn print_and_format_contract_info(info: ContractInfo) {
+        name_value_println!(
+            "TrieId for the substree",
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/extrinsics/info.rs:105]:
```suggestion
            "TrieId:",
```

File: crates/cargo-contract/src/cmd/extrinsics/info.rs [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+            format!("{:?}", info.trie_id),
+            DEFAULT_KEY_COL_WIDTH
+        );
+        name_value_println!(
+            "Code hash for the given account",
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/extrinsics/info.rs:110]:
```suggestion
            "Code hash:",
```

File: crates/cargo-contract/src/cmd/extrinsics/info.rs [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+            format!("{:?}", info.code_hash),
+            DEFAULT_KEY_COL_WIDTH
+        );
+        name_value_println!(
+            "Items of storage accumulated in this contract",
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/extrinsics/info.rs:115]:
```suggestion
            "Number of storage items:",
```

File: crates/cargo-contract/src/cmd/extrinsics/info.rs [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+            format!("{:?}", info.storage_items),
+            DEFAULT_KEY_COL_WIDTH
+        );
+        name_value_println!(
+            "How much deposit the accumulated `storage_items` amount to",
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/extrinsics/info.rs:120]:
```suggestion
            "Storage deposit:",
```

File: crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs [hunk 11]
```diff
@@ -309,3 +309,94 @@ async fn build_upload_remove() {
...
+///
+/// # Note
+///
+/// Requires [substrate-contracts-node](https://github.com/paritytech/substrate-contracts-node/) to
+/// be installed and available on the `PATH`, and the no other process running using the default
+/// port `9944`.
+///
+/// ** This test is ignored for now since the substrate-contracts-node is not installed on CI **
+/// It will be addressed in a follow up PR, for now it can be run locally by commenting out the
+/// `ignore` attribute below
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs:324]: There is no more `ignore` attribute, so I guess the comment can be removed?

File: crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs [hunk 12]
```diff
@@ -309,3 +309,94 @@ async fn build_upload_remove() {
...
+
+    // Spawn the contracts node
+    let node_process = ContractsNodeProcess::spawn(CONTRACTS_NODE)
+        .await
+        .expect("Error spawning contracts node");
+
+    tracing::debug!(
+        "Creating new contract in temporary directory {}",
+        tmp_dir.path().to_string_lossy()
+    );
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs:342]:
```suggestion
```

For tests, we don't include tracing output. It's fine to have this during development or for debugging, but shouldn't be in `master`.

File: crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs [hunk 13]
```diff
@@ -309,3 +309,94 @@ async fn build_upload_remove() {
...
+#[async_std::test]
+async fn build_upload_instantiate_info() {
+    init_tracing_subscriber();
+
+    let tmp_dir = tempfile::Builder::new()
+        .prefix("cargo-contract.cli.test.")
+        .tempdir()
+        .expect("temporary directory creation failed");
+
+    // Spawn the contracts node
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs:334]:
```suggestion
```

File: crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs [hunk 14]
```diff
@@ -309,3 +309,94 @@ async fn build_upload_remove() {
...
+    let node_process = ContractsNodeProcess::spawn(CONTRACTS_NODE)
+        .await
+        .expect("Error spawning contracts node");
+
+    tracing::debug!(
+        "Creating new contract in temporary directory {}",
+        tmp_dir.path().to_string_lossy()
+    );
+
+    // cargo contract new flipper
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs:344]:
```suggestion
```

File: README.md (hunk 1 above)
Comment (lean-apple) [reply README.md]: Agree, was moved out

File: crates/cargo-contract/src/cmd/info.rs [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+            let info_result = self.fetch_contract_info(&client).await?;
+
+            match info_result {
+                Some(info_result) => {
+                    InfoCommand::print_and_format_contract_info(info_result);
+                }
+                None => {
+                    return Err(anyhow!(
+                        "No contract information was found for the ContractId {}",
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/info.rs:81]:
```suggestion
                        "No information was found for account id {}",
```

File: crates/cargo-contract/src/cmd/info.rs [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+impl InfoCommand {
+    pub fn is_json(&self) -> bool {
+        self.output_json
+    }
+
+    pub fn run(&self) -> Result<(), ErrorVariant> {
+        tracing::debug!(
+            "Getting information for contract AccountId {:?}",
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/info.rs:65]:
```suggestion
            "Getting contract information for AccountId {:?}",
```

File: crates/cargo-contract/src/main.rs [hunk 17]
```diff
@@ -130,6 +131,9 @@ enum Command {
     /// Remove contract code
     #[clap(name = "remove")]
     Remove(RemoveCommand),
+    /// Get info contract
```
Comment (cmichi) [inline crates/cargo-contract/src/main.rs:134]:
```suggestion
    /// Display contract information
```

File: crates/cargo-contract/src/cmd/info.rs [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+use std::fmt::Debug;
+use subxt::{
+    Config,
+    OnlineClient,
+};
+
+#[derive(Debug, clap::Args)]
+#[clap(name = "info", about = "Get infos from a contract")]
+pub struct InfoCommand {
+    /// The address of the the contract to call.
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/info.rs:43]:
```suggestion
    /// Address of the contract to call.
```

File: crates/cargo-contract/src/cmd/info.rs [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+    #[clap(name = "contract", long, env = "CONTRACT")]
+    contract: <DefaultConfig as Config>::AccountId,
+    #[clap(
+        name = "url",
+        long,
+        value_parser,
+        default_value = "ws://localhost:9944"
+    )]
+    url: url::Url,
+    /// Export the call output in JSON format.
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/info.rs:53]:
```suggestion
    /// Export the info output as JSON.
```

**Comment by cmichi** [comment]
I'm currently getting 
```bash
$ cargo run -- contract info --contract 5HdqKDLPe2VgUZ3v1kGu28yEJNzE7rSSVpYNXWfXvDivfEML              
    Finished dev [unoptimized + debuginfo] target(s) in 0.21s
     Running `target/debug/cargo-contract contract info --contract 5HdqKDLPe2VgUZ3v1kGu28yEJNzE7rSSVpYNXWfXvDivfEML`
thread 'main' panicked at 'Command info: Argument or group 'verbose' specified in 'conflicts_with*' for 'output_json' does not exist', /Users/michi/.cargo/registry/src/github.com-1ecc6299db9ec823/clap-4.1.8/src/builder/debug_asserts.rs:217:13
```

**Comment by lean-apple** [comment]
In fact, is this flag `output_json` useful at the moment, in the state of this command  and in the way it is used now ?

File: crates/cargo-contract/src/main.rs [hunk 20]
```diff
@@ -130,6 +131,9 @@ enum Command {
     /// Remove contract code
     #[clap(name = "remove")]
     Remove(RemoveCommand),
+    /// Display contract information
```
Comment (cmichi) [inline crates/cargo-contract/src/main.rs:134]:
```suggestion
    /// Display information about a contract
```

File: crates/cargo-contract/src/cmd/info.rs [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+use std::fmt::Debug;
+use subxt::{
+    Config,
+    OnlineClient,
+};
+
+#[derive(Debug, clap::Args)]
+#[clap(name = "info", about = "Get infos from a contract")]
+pub struct InfoCommand {
+    /// Address of the contract to display info.
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/info.rs:43]:
```suggestion
    /// The address of the contract to display info of.
```

File: docs/info.md [hunk 22]
```diff
@@ -0,0 +1,23 @@
+# Other commands
+`cargo-contract` provides CLI support for displaying info directly from the command
+line.
+
+### `info`
+
+Fetch and display informations for a given contractId (AccountId).
```
Comment (cmichi) [inline docs/info.md:7]:
```suggestion
Fetch and display information for a given contract.
```

File: docs/info.md [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+### `info`
+
+Fetch and display informations for a given contractId (AccountId). 
+Display these information (https://github.com/paritytech/substrate/blob/master/frame/contracts/src/storage.rs#L45-L69) from 
+
+
+e.g. `cargo contract info --suri //Alice`
+
+Assumes that `cargo contract build`, `cargo contract upload` and `cargo contract instantiate` have already been run to produce the contract artifacts.
+
```
Comment (cmichi) [inline docs/info.md:14]:
```suggestion
```

This is not correct, also `--suri` is not necessary for `cargo contract info`.

File: docs/info.md [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+```
+cargo contract info \
+       --contract 5DVGLfDGBvqMr9nCg48g99oD8Mz3sruWmb6ek5UbWvDnbTgZ
+```
+
+- `--contract` the account id of the contract to invoke, returned after a successful `contract instantiate`
+
+*Optional*
+- `--url` the url of the rpc endpoint you want to specify - by default ws://localhost:9944 .
```
Comment (cmichi) [inline docs/info.md:23]:
```suggestion
- `--url` the url of the rpc endpoint you want to specify - by default `ws://localhost:9944`.
```

File: crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs [hunk 25]
```diff
@@ -309,3 +293,81 @@ async fn build_upload_remove() {
...
+    // cd flipper
+    let mut project_path = tmp_dir.path().to_path_buf();
+    project_path.push("flipper");
+
+    cargo_contract(project_path.as_path())
+        .arg("build")
+        .assert()
+        .success();
+
+    tracing::debug!("Uploading the code to the substrate-contracts-node chain");
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs:333]:
```suggestion
    tracing::debug!("Uploading contract to substrate-contracts-node");
```

File: crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs [hunk 26]
```diff
@@ -309,3 +293,81 @@ async fn build_upload_remove() {
     // prevent the node_process from being dropped and killed
     let _ = node_process;
 }
+
+/// Sanity test the whole lifecycle of:
+///   new -> build -> upload -> instantiate -> info
+///
+/// # Note
+///
+/// Requires [substrate-contracts-node](https://github.com/paritytech/substrate-contracts-node/) to
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs:302]:
```suggestion
/// Requires [`substrate-contracts-node`](https://github.com/paritytech/substrate-contracts-node/) to
```

File: crates/cargo-contract/src/cmd/info.rs [hunk 27]
```diff
@@ -0,0 +1,121 @@
...
+    fn print_and_format_contract_info(info: ContractInfo) {
+        let convert_trie_id = hex::encode(info.trie_id.0);
+        name_value_println!(
+            "TrieId:",
+            format!("{}", convert_trie_id),
+            DEFAULT_KEY_COL_WIDTH
+        );
+        name_value_println!(
+            "Code hash:",
+            format!("{}", info.code_hash),
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/info.rs:107]: image

This is what I'm getting, it should be the unshortened hash here.

File: docs/info.md [hunk 28]
```diff
@@ -0,0 +1,23 @@
...
+e.g. `cargo contract info --suri //Alice`
+
+Assumes that `cargo contract build`, `cargo contract upload` and `cargo contract instantiate` have already been run to produce the contract artifacts.
+
+```
+cargo contract info \
+       --contract 5DVGLfDGBvqMr9nCg48g99oD8Mz3sruWmb6ek5UbWvDnbTgZ
+```
+
+- `--contract` the account id of the contract to invoke, returned after a successful `contract instantiate`
```
[the remark this answers is not in the document]
Comment (lean-apple) [reply docs/info.md:20]: What is the purpose of invoking `cargo contract info` if the contract is instantiated ?

[the remark this answers is not in the document]
Comment (cmichi) [reply docs/info.md]: I removed

>returned after a successful `contract instantiate`

because doing `cargo contract instantiate` is not needed to be done before `cargo contract info`. A user can also obtain a contract id via some other mean.

[the remark this answers is not in the document]
Comment (lean-apple) [reply docs/info.md]: Yes I agree on the fact that the user can get the account id in other way but if these commands are not run, no information will be displayed.

[the remark this answers is not in the document]
Comment (cmichi) [reply docs/info.md]: Executing those commands is not the only way to deploy a contract, users can also use e.g. the Contracts UI, `polkadot-js` or Swanky.

File: crates/cargo-contract/src/cmd/extrinsics/info.rs (hunk 3 above)
Comment (lean-apple) [reply crates/cargo-contract/src/cmd/extrinsics/info.rs]: Yes, I couldnt find a way to not use it. The other alternative is to move it to the Some block.

Comment (ascjones) [reply crates/cargo-contract/src/cmd/extrinsics/info.rs]: I just tried with `Ok(())` and it compiled successfully.

File: crates/cargo-contract/src/cmd/extrinsics/upload.rs [hunk 29]
```diff
@@ -158,7 +160,7 @@ impl UploadCommand {
         let token_metadata = TokenMetadata::query(client).await?;
         let storage_deposit_limit =
             self.extrinsic_opts.storage_deposit_limit(&token_metadata)?;
-        let call = super::runtime_api::api::tx().contracts().upload_code(
+        let call = crate::cmd::runtime_api::api::tx().contracts().upload_code(
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/extrinsics/upload.rs:163]: What is the reason for this change?

File: crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs [hunk 30]
```diff
@@ -309,3 +277,77 @@ async fn build_upload_remove() {
...
+    let stdout = str::from_utf8(&output.stdout).unwrap();
+    let stderr = str::from_utf8(&output.stderr).unwrap();
+    assert!(output.status.success(), "instantiate failed: {stderr}");
+
+    // find the contract address in the output
+    let regex = regex::Regex::new("Contract ([0-9A-Za-z]+)").unwrap();
+    let caps = regex
+        .captures(stdout)
+        .expect("contract account regex capture");
+    let contract_account = caps.get(1).unwrap().as_str();
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs:340]: Please extract into a helper function `fn extract_contract_address(stdout)`. You can then also remove the code duplication in `build_upload_instantiate_call`.

File: crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs [hunk 31]
```diff
@@ -110,7 +105,6 @@ impl ContractsNodeProcess {
     }
 
     fn kill(&mut self) {
-        tracing::debug!("Killing contracts node process {}", self.proc.id());
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs:113]: As this one is in a helper function and not directly in the test, let's bring it back.

File: crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs [hunk 32]
```diff
@@ -71,11 +71,6 @@ impl ContractsNodeProcess {
         let mut attempts = 1;
         let client = loop {
             thread::sleep(time::Duration::from_secs(1));
-            tracing::debug!(
-                "Connecting to contracts enabled node, attempt {}/{}",
-                attempts,
-                MAX_ATTEMPTS
-            );
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/extrinsics/integration_tests.rs:78]: As this one is in a helper function and not directly in the test, let's bring it back.

File: crates/cargo-contract/src/cmd/info.rs [hunk 33]
```diff
@@ -0,0 +1,107 @@
...
+    OnlineClient,
+};
+
+#[derive(Debug, clap::Args)]
+#[clap(name = "info", about = "Get infos from a contract")]
+pub struct InfoCommand {
+    /// The address of the contract to display info of.
+    #[clap(name = "contract", long, env = "CONTRACT")]
+    contract: <DefaultConfig as Config>::AccountId,
+    #[clap(
```
Comment (cmichi) [inline crates/cargo-contract/src/cmd/info.rs:45]:
```suggestion
    /// Websockets url of a substrate node.
    #[clap(
```

File: CHANGELOG.md [hunk 34]
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 - Add `cargo contract encode` command - [#998](https://github.com/paritytech/cargo-contract/pull/998)
+- Add `info` command - [#993](https://github.com/paritytech/cargo-contract/pull/993)
```
Comment (cmichi) [inline CHANGELOG.md:14]: Please merge `master`into your PR, you'll have to put this into the unreleased section again then.

**Review by cmichi — APPROVED** [review]
Thank you, Lea!
